homeassistant/components/mqtt/fan.py

```diff
@@ -217,19 +354,107 @@ def state_received(msg):
                 "qos": self._config[CONF_QOS],
             }
 
+        @callback
+        @log_messages(self.hass, self.entity_id)
+        def percentage_received(msg):
+            """Handle new received MQTT message for the percentage."""
+            numeric_val_str = self._templates[ATTR_PERCENTAGE](msg.payload)
+            try:
+                percentage = ranged_value_to_percentage(
+                    self._speed_range, int(numeric_val_str)
+                )
+            except ValueError:
+                _LOGGER.warning(
+                    "'%s' received on topic %s is not a valid speed within the speed range",
+                    msg.payload,
+                    msg.topic,
+                )
+                return
+            if 0 <= percentage <= 100:
+                self._percentage = percentage
+            else:
+                _LOGGER.warning(
+                    "'%s' received on topic %s is not a valid speed within the speed range",
+                    msg.payload,
+                    msg.topic,
+                )
+                return
+            self.async_write_ha_state()
```

Comment:
Use a guard clause instead:
```suggestion
            if percentage < 0 or percentage > 100:
                _LOGGER.warning(
                    "'%s' received on topic %s is not a valid speed within the speed range",
                    msg.payload,
                    msg.topic,
                )
                return
            self.async_write_ha_state()
```